testing/src/scenario/context.py

```diff
@@ -365,6 +365,79 @@ def action(
             kwargs["id"] = id
         return _Event(f"{name}_action", action=_Action(name, **kwargs))
 
+    @staticmethod
+    def custom(
+        source: ops.EventSource,
+        *,
+        args: tuple[Any] | None = None,
+        kwargs: dict[str, Any] | None = None,
+        from_juju_event: _Event | None = None,
+    ):
+        """Event triggered by a charm library.
+
+        For example, for a library that emits a ``HappenedEvent`` event via a
+        ``MyConsumer`` object, and charm code like::
+
+            class MyCharm(ops.CharmBase):
+                def __init__(self, framework: ops.Framework):
+                    super().__init__(framework)
+                    self.source = MyConsumer(self, 'source')
+                    framework.observe(self.source.on.it_happened, self._on_happened_event)
+
+        Emit this event with::
+
+            ctx.run(ctx.on.custom(
+                MyConsumer.on.it_happened, args=(1, 2), kwargs={'foo': 'bar'}
+            ), state)
+
+        If the custom event is a subclass of a Juju event (rather than :attr:`ops.EventBase`)
+        that itself has arguments (such as :attr:`ops.RelationChangedEvent`),
+        then you need to also pass in a suitable originating event, for example::
+
+            ctx.run(ctx.on.custom(
+                MyConsumer.on.it_happened, from_juju_event=ctx.on.relation_changed(my_relation)
+            ), state)
+
+        The arguments for the custom event are the arguments for the original
+        Juju event, combined with any that are passed here.
+        """
+        event_kwargs = {}
+        if from_juju_event:
+            name = from_juju_event.name
+            if from_juju_event.storage is not None:
+                event_kwargs["storage"] = from_juju_event.storage
+            if from_juju_event.relation is not None:
+                event_kwargs["relation"] = from_juju_event.relation
```

Comment:
It would be weird to have both set, wouldn't it?

Reply:
Yeah, most of these are actually groups, where it makes no sense to get both of them. It felt simpler in this code to just copy everything that was set. I could instead add an `_Event.merge` or something like that, which would look at the type of event and only copy the relevant ones, or raise an exception if it was inconsistent. Or `_Event` could itself have a `__post_init__` that verified consistency? Or maybe there should be `_Event` subclasses for the different types rather than the catch-all it currently is?

Not sure if it's worth doing any of these in this PR, wdyt?

Reply:
No longer relevant, since this piece of the feature was removed.